`QueryManager.inFlightLinkObservables`: use a strong `Trie`

phryneas opened the pull request:
Since `@wryware/trie@0.5.0` now has a `remove` function, we can use that data structure in `QueryManager.inFlightLinkObservables` and shave off a few bytes here.

## Should not be merged before we have a new `optimism` version
## Before merging/releasing: add handling to `@apollo/experimental-nextjs-app-support`


### Checklist:

- [x] If this PR contains changes to the library itself (not necessary for e.g. docs updates), please include a changeset (see [CONTRIBUTING.md](https://github.com/apollographql/apollo-client/blob/main/CONTRIBUTING.md#changesets))
- [ ] If this PR is a new feature, please reference an issue where a consensus about the design was reached (not necessary for small changes)
- [ ] Make sure all of the significant new logic is covered by tests

jerelmiller commented:
@phryneas just to clarify the intent of this change, is this strictly to shave off some bytes from the end build size, or will this have a memory impact as well since the printed query string can now be shared for multiple variables in the `Trie`? Would love to have a quick "why" in the changeset to describe the impact of this change so end users understand how this will positively impact them when upgrading. Thanks!

phryneas commented:
@jerelmiller This is really just a "shaving off bytes" - the structure was a Map of Maps before and is a Map of Maps (in the form of a Trie) now. Thing is, we add slightly more bundle size as the `Trie` implementation now has a `remove` method, so we can try to counteract that by using the `Trie` implementation where it saves a bit of code - and here, we essentially had a hand-written `Trie` in our sourcecode.

=> Implementation changes, but no observable behaviour will change in any way. That will happen in other places where we will use the new `remove` method (e.g. `SuspenseCache`).

jerelmiller commented:
@phryneas thats helpful to understand! Appreciate the context here :)

phryneas commented:
/release:pr

phryneas commented:
/release:pr

phryneas commented:
I've fixed up some tests, added a runtime warning if used with oudated versions of `@apollo/experimental-nextjs-app-support`, and had some minor code changes to ensure compatibility.

Requesting a re-review for that reason :)

You can see these changes here:

https://github.com/apollographql/apollo-client//compare/4d6e8c3c0421c2c8ed268c6cb4231948c195dd0f...0cbdfcc71e8c9edba3e9d7eca1e5548435beee07

On the other side, I've opened the PR https://github.com/apollographql/apollo-client-nextjs/pull/144 to accomodate for the changes in this PR. Please give it a review, too :)

phryneas commented on src/core/QueryManager.ts:
```diff
@@ -182,6 +183,15 @@ export class QueryManager<TStore> {
       this.mutationStore = Object.create(null);
     }
+
+    // TODO: remove before we release 3.9
+    Object.defineProperty(this.inFlightLinkObservables, "get", {
+      value: () => {
+        throw new Error(
+          "This version of Apollo Client requires at least @apollo/experimental-nextjs-app-support version 0.5.2."
+        );
+      },
+    });
```
You can see this work out over in https://github.com/apollographql/apollo-client-nextjs/pull/144#issuecomment-1833826751

jerelmiller reviewed (approved):
Thanks for catching that update to the Next.js package!
